src/quick-lint-js/parse.h

```diff
@@ -2460,6 +2460,71 @@ class parser {
     v.visit_exit_with_scope();
   }
 
+  bool is_side_effect_free(expression *ast) {
```

Comment:
I think this function could easily be tested. Let me know if you want some help writing tests. (I think the tests would go in test-parse.cpp or test-parse-expression.cpp.) I provided some non-obvious/tricky examples in some comments.